PowerShell/JumpCloud Module/Public/Directory/Add-JCGsuiteMember.ps1

```diff
@@ -0,0 +1,106 @@
+function Add-JCGsuiteMember () {
+    param(
+        [Parameter( ValueFromPipelineByPropertyName, ParameterSetName = 'ByName', HelpMessage = 'The name of cloud directory instance')]
+        [String]$Name,
+        [Parameter( ValueFromPipelineByPropertyName, ParameterSetName = 'ByID', HelpMessage = 'The ID of cloud directory instance')]
+        [Alias('_id')]
+        [String]$ID,
+        [Parameter( ValueFromPipelineByPropertyName, HelpMessage = 'A username to add to the directory')]
+        [String]$Username,
+        [Parameter( ValueFromPipelineByPropertyName, HelpMessage = 'A UserID to add to the directory')]
+        [String]$UserID,
+        [Parameter( ValueFromPipelineByPropertyName, HelpMessage = 'A UserGroup ID to add to the directory')]
+        [String]$GroupID,
+        [Parameter( ValueFromPipelineByPropertyName, HelpMessage = 'A UserGroup name to add to the directory')]
+        [String]$GroupName
+    )
+    begin {
+        Write-Debug 'Verifying JCAPI Key'
+        if ($JCAPIKEY.length -ne 40) {
+            Connect-JCOnline
+        }
+        $resultsArray = [System.Collections.Generic.List[PSObject]]::new()
+        $DirectoryHash = Get-JcSdkDirectory | Where-Object type -EQ 'g_suite' | Select-Object id, name
+        if (($Username -or $UserID) -and ($GroupID -or $GroupName)) {
+            throw "Please add one type of association per call"
+        } elseif ($Username -and $UserID) {
+            throw "Please use either a username or a userID"
+        } elseif ($GroupID -and $GroupName) {
+            throw "Please use either a UserGroup Name or a UserGroup ID"
+        } elseif ($Username -or $UserID) {
+            Write-Debug 'Populating UserHash'
+            $UserHash = Get-DynamicHash -Object User -returnProperties username
+        } elseif ($GroupID -or $GroupName) {
+            Write-Debug 'Populating UserGroupHash'
+            $UserGroupHash = Get-DynamicHash -Object Group -GroupType User -returnProperties name
```

Comment:
Just wondering about the efficiency here. Technically speaking, if we are going to add a list of users to a cloud directory object. This doesn't feel like the best way for us to be doing this. 

Imagine a list of 10 users we want to add to some Google Workspace Directory:
User_0
User_1
...
User_9

and we write a loop to do this, each user is added with 
`Add-JCGSuiteMember -Name "GoogleDir" -Username "User_0"`

We'd be getting **A LOT** of data somewhat unnecessarily. I know that the DynamicHash functions are quick but it's a lot of data to gather each iteration for just finding one user. I'd really argue that we are better off with the search functions here. i.e.
If Username:
Get-JCUser -username "User_0" -returnProperties username
If UserID:
Get-JCUser -userID "12345" -returnProperties username

We'd still be giving the end user a warning if the username or userID was not found but in this case I'd really argue that getting all users is not performant.

I think we'd want to do the same thing with groups.

Reply:
+1 on Joe's point, we can maybe be more efficient by not getting all user/group objects if we are only pulling one or two data.